server/src/core/api-rest/api-rest-query-builder/factories/parsers/filter-parser.factory.ts

```diff
@@ -0,0 +1,216 @@
+import { Injectable } from '@nestjs/common';
+
+import { Request } from 'express';
+
+import {
+  checkFields,
+  getFieldType,
+} from 'src/core/api-rest/utils/metadata-query.utils';
+
+enum FILTER_COMPARATORS {
+  eq = 'eq',
+  neq = 'neq',
+  in = 'in',
+  is = 'is',
+  gt = 'gt',
+  gte = 'gte',
+  lt = 'lt',
+  lte = 'lte',
+  startsWith = 'startsWith',
+  like = 'like',
+  ilike = 'ilike',
+}
+
+enum CONJUNCTIONS {
+  or = 'or',
+  and = 'and',
+  not = 'not',
+}
+
+@Injectable()
+export class FilterParserFactory {
+  create(request: Request, objectMetadata) {
+    let filterQuery = request.query.filter;
+    if (typeof filterQuery !== 'string') {
+      return {};
+    }
+    this.checkFilterQuery(filterQuery);
+    filterQuery = this.addDefaultConjunctionIfMissing(filterQuery);
+    return this.parseStringFilter(
+      filterQuery,
+      objectMetadata.objectMetadataItem,
+    );
+  }
+
+  addDefaultConjunctionIfMissing(filterQuery) {
+    if (!(filterQuery.includes('(') && filterQuery.includes(')'))) {
+      return `and(${filterQuery})`;
+    }
+    return filterQuery;
+  }
+
+  checkFilterQuery(filterQuery) {
+    const countOpenedBrackets = (filterQuery.match(/\(/g) || []).length;
+    const countClosedBrackets = (filterQuery.match(/\)/g) || []).length;
+    const diff = countOpenedBrackets - countClosedBrackets;
+    if (diff !== 0) {
+      const hint =
+        diff > 0
+          ? `${diff} close bracket${diff > 1 ? 's are' : ' is'}`
+          : `${Math.abs(diff)} open bracket${
+              Math.abs(diff) > 1 ? 's are' : ' is'
+            }`;
+      throw Error(`'filter' invalid. ${hint} missing in the query`);
+    }
+    return;
+  }
+
+  parseFilterQueryContent(filterQuery) {
+    let bracketsCounter = 0;
+    let parenthesisCounter = 0;
+    const predicates: string[] = [];
+    let currentPredicates = '';
+    for (const c of filterQuery) {
+      if (c === '(') {
+        parenthesisCounter++;
+        if (parenthesisCounter === 1) continue;
+      }
+      if (c === ')') {
+        parenthesisCounter--;
+        if (parenthesisCounter === 0) continue;
+      }
+      if (c === '[') {
+        bracketsCounter++;
+      }
+      if (c === ']') {
+        bracketsCounter--;
+      }
+      if (c === ',' && parenthesisCounter === 1 && bracketsCounter === 0) {
+        predicates.push(currentPredicates);
+        currentPredicates = '';
+        continue;
+      }
+      if (parenthesisCounter >= 1) currentPredicates += c;
+    }
+    if (currentPredicates.length) {
+      predicates.push(currentPredicates);
+    }
+    return predicates;
+  }
+
+  parseSimpleFilterString(filterString: string): {
```

Comment:
@martmull Same not really easy to understand what this function is doing, adding a comment will be helpful